src/ImageSharp/Image.FromStream.cs

```diff
@@ -574,34 +583,36 @@ public static Image Load(DecoderOptions options, Stream stream, out IImageFormat
             throw new NotSupportedException("Cannot read from the stream.");
         }
 
-        Configuration configuration = options.Configuration;
-        if (stream.CanSeek && configuration.ReadOrigin == ReadOrigin.Begin)
+        async Task<T> Action(Stream s, long position, CancellationToken ct)
         {
-            stream.Position = 0;
+            T result = await action(s, ct).ConfigureAwait(false);
 
-            // NOTE: We are explicitly not executing the action against the stream here as we do in WithSeekableStream() because that
-            // would incur synchronous IO reads which must be avoided in this asynchronous method.  Instead, we will *always* run the
-            // code below to copy the stream to an in-memory buffer before invoking the action.
-        }
+            // Our buffered reads may have left the stream in an incorrect non-zero position.
+            // Reset the position of the seekable stream if we did not read to the end to allow additional reads.
+            if (stream.CanSeek && stream.Position != s.Position && s.Position != s.Length)
+            {
+                stream.Position = position + s.Position;
+            }
 
-        using ChunkedMemoryStream memoryStream = new(configuration.MemoryAllocator);
-        await stream.CopyToAsync(memoryStream, configuration.StreamProcessingBufferSize, cancellationToken).ConfigureAwait(false);
-        memoryStream.Position = 0;
+            return result;
+        }
 
-        T Action(Stream ms, CancellationToken ct)
+        Configuration configuration = options.Configuration;
+        if (stream.CanSeek)
         {
-            // Reset the position of the seekable stream if we did not read to the end
-            // to allow additional reads.
-            T result = action(ms, ct);
-            if (stream.CanSeek && ms.Position != ms.Length)
+            if (configuration.ReadOrigin == ReadOrigin.Begin)
             {
-                stream.Position = ms.Position;
+                stream.Position = 0;
             }
 
-            return result;
+            return await Action(stream, stream.Position, cancellationToken).ConfigureAwait(false);
```

Comment:
We now allow external implementations to be asynchronous and only copy when necessary.